sql/catalyst/src/test/scala/org/apache/spark/sql/types/TestUDT.scala

```diff
@@ -58,4 +63,22 @@ object TestUDT {
 
     override def equals(other: Any): Boolean = other.isInstanceOf[MyDenseVectorUDT]
   }
+
+  private[sql] class MyXMLGregorianCalendarUDT extends UserDefinedType[XMLGregorianCalendar] {
+    override def sqlType: DataType = TimestampType
+
+    override def serialize(obj: XMLGregorianCalendar): Any =
+      obj.toGregorianCalendar.getTimeInMillis * 1000
+
+    override def deserialize(datum: Any): XMLGregorianCalendar = {
+      val calendar = new GregorianCalendar
+      calendar.setTimeInMillis(datum.asInstanceOf[Long])
+      DatatypeFactory.newInstance.newXMLGregorianCalendar(calendar)
+    }
+
+    override def userClass: Class[XMLGregorianCalendar] = classOf[XMLGregorianCalendar]
+
+    // By setting this to a timestamp, we lose the information about the udt
+    override private[sql] def jsonValue: JValue = "timestamp"
```

Comment:
Why did you change this? ISTM this method is not expected to be override by users. We need `final` for `UserDefinedType.jsonValue`? cc: @dongjoon-hyun @HyukjinKwon 
 https://github.com/apache/spark/blob/2a28c73d81dd12fa39eb52d3eb343e651123a51d/sql/catalyst/src/main/scala/org/apache/spark/sql/types/UserDefinedType.scala#L61

Reply:
With the normal UDT `jsonValue`, we would get:
```
  override private[sql] def jsonValue: JValue = {
    ("type" -> "udt") ~
      ("class" -> this.getClass.getName) ~
      ("pyClass" -> pyUDT) ~
      ("sqlType" -> sqlType.jsonValue)
```
Which will write the type as the UDT. If you try to read the column later on in another job where the UDTRegistration hasn't been done will give an error. Therefore we would like to write the XMLGregorianCalendar as a normal timestamp. However, when we append the table, we want to be able to merge the `XMLGregorianCalendar` UDT into the timestamp. Without the added rule this wouldn't be possible. Hope this helps.

Reply:
This is a bit odd .. wouldn't it be a timestamp whenever it does a roundtrip in ser/de?

Reply:
That is correct. So when we serialize it, it will use:
```scala
    override def serialize(obj: XMLGregorianCalendar): Any =
      obj.toGregorianCalendar.getTimeInMillis * 1000
```
Which will write it as a timestamp stored as the number of microseconds from the epoch of 1970-01-01T00:00:00.000000Z (UTC+00:00).

And this serialize implementation should be the same as the one in the `sqlType`, because we don't store any references to the UDT, another Spark session will just decode it as a `TimestampType`, so the decoder of the TimestampType will be used. 

I know that this isn't trivial, but it is very powerful and makes the UDT's much more flexible when writing ETL jobs since you can directly use your own types with the Dataset API.